pyanaconda/kickstart_dispatcher.py

```diff
@@ -0,0 +1,355 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+from collections import namedtuple
+
+from pykickstart.parser import KickstartParser
+from pykickstart.sections import Section
+
+KickstartElement = namedtuple('KickstartElement',
+                              ['args', 'content', 'lineno', 'filename'])
+
+
+class KickstartElements(object):
+    """Container for kickstart elements (commands, sections, addons)
+
+    Preserves order of added elements. Allows filtering of elements.
+    """
+
+    def __init__(self):
+        self._elements = []
+
+    def append_command(self, name, content, lineno, filename):
+        """Appends kickstart command to the container.
+
+        :param name: name of the command
+        :type name: str
+        :param content: command line including newline
+        :type content: str
+        :param lineno: line number of the command in a kickstart file
+        :type content: int
+        :param filename: name of a kickstart file
+        :type filename: str
+        """
+        command = KickstartElement(args=(name,),
+                                   content=content,
+                                   lineno=lineno,
+                                   filename=filename)
+        self._elements.append(command)
+
+    def append_section(self, name, content, header_lineno, filename):
+        """Appends kickstart section to the container.
+
+        :param name: name of the section (including leading '%' char)
+        :type name: str
+        :param content: section content (ending with newline)
+        :type content: str
+        :param header_lineno: line number of the section header in a kickstart file
+        :type content: int
+        :param filename: name of a kickstart file
+        :type filename: str
+        """
+        section = KickstartElement(args=(name,),
+                                   content=content,
+                                   lineno=header_lineno,
+                                   filename=filename)
+        self._elements.append(section)
+
+    def append_addon(self, name, content, header_lineno, filename):
+        """Appends addon kickstart section to the container.
+
+        :param name: name of the addon
+        :type name: str
+        :param content: addon section content (ending with newline)
+        :type content: str
+        :param header_lineno: line number of the addon section header in a kickstart file
+        :type content: int
+        :param filename: name of a kickstart file
+        :type filename: str
+        """
+        addon = KickstartElement(args=("%addon", name),
+                                 content=content,
+                                 lineno=header_lineno,
+                                 filename=filename)
+        self._elements.append(addon)
```

Comment:
I'll try this approach, I was thinking about having a class for element. Then it could/should also take care also of tracking (by setting a flag on the instance). Now the tracking takes advantage to element being immutable, so it is possible to use set for tracking processed elements in TrackedKickstartElements.